### PR TITLE
February 2021 update

### LINKED ISSUE #219: Feature Request: Add the Control Pictures Unicode block
We talked about control characters before, and how they are interpreted by the console or the terminal instead of printing characters.
When working in a terminal, it is sometimes helpful to visualize these in-band control sequences, Visual Studio Code even does it when showing files, by showing small "ESC", "SUB",... when the "Render Control Characters" option is enabled.

Unicode got the same idea, and included a block of Control Pictures (U+2400 to U+2426) in Unicode 12.0. These are designed to be able to represent the control characters on a terminal screen:
␀␁␂␃␄␅␆␇␈␉␊␋␌␍␎␏␐␑␒␓␔␕␖␗␘␙␚␛␜␝␞␟␠␡␢␣␤␥␦
(https://en.wikipedia.org/wiki/Control_Pictures)
When available, these can be used by CUI apps to provide a visual representation of these special characters, exactly like Visual Studio Code does.

Adding these 39 glyphs would make it possible for utilities such as hexdump to show them in text representations, which is much more helpful than having 34 of the 256 values show up as generic dots. It would even make it possible to a CUI text editor to provide the "Render Control Characters" option.

Windows Terminal currently falls back to another font to render these, but they are tiny and impractical for use in a terminal.

Below is a sample of a hexdump function showing the contents of cmd.exe with high-ascii and control characters (using the font fallback):
hexdump with Control Pictures
And the Ubuntu hexdump command showing the same file, with dots for high-ascii and control characters (so only 96 values out of 256 provide chars representations). (this one is not using Cascadia, but shows the limitation when these characters are not available).
hexdump without Control Pictures

### LINKED ISSUE #262: Add support for U+23CE
# Description of the new feature

Add support for ⏎ (U+23CE "RETURN SYMBOL")

Currently it looks like this: 
cascadia-return-symbol-missing

## Why?

`fish` uses ⏎ to indicate line breaks added to preserve prompt structure

### LINKED ISSUE #264: Feature Request: Code points for control/hidden chars (e.g. U+21B2, U+2771, U+2770, U+2423)
Hi guys, thank you very much for Cascadia, it is really good.

The usage of codepoints (e.g. U+21B2, U+2771, U+2770, U+2423) to represent hidden/control chars is [common pratice in text editors](https://www.reddit.com/r/vim/comments/4hoa6e/what_do_you_use_for_your_listchars/)  to provide extra information about the text and disambiguation. Particularly, I have the following in my vimrc:
```vim
set listchars=tab:▶\ ,trail:•,extends:❱,precedes:❰,nbsp:␣,eol:↲
```
This is how it is rendered right now (FVim front end that uses [Avalonia](http://avaloniaui.net) and F#):
image
This is more or less how it could be
image

I was wondering if we could have something like these codepoints.

### LINKED ISSUE #276: Incorporate new ligatures in Fira Code v2, v3, v4 and v5
Fira Code v3 was released last month and v4 this week. v2 & v3 includes a large number of new coding ligatures (from the looks of things the ligatures from Fira Code seem to be before v2).

Releases show the added ligatures and what they look like:
https://github.com/tonsky/FiraCode/releases/tag/4
https://github.com/tonsky/FiraCode/releases/tag/3
https://github.com/tonsky/FiraCode/releases/tag/2

v4 features infinite arrow combinations instead of tens of individual ligatures:

### LINKED ISSUE #281: Feature Request: Add !: and !. TypeScript ligatures
Considering that `?:` and `?.` ligatures are already there, I think that adding these will make the code look more consistent.

### LINKED ISSUE #290: Add ligature for /\ and \/
Hello,
Please add ligature for `/\` and `\/`
Like "Fira Code" font

### LINKED ISSUE #301: Feature Request: `??=` ligature
# Description of the new feature

There’s already a ligature for `?=`, `??` and `||=`, but not one for `??=`.

# Proposed technical implementation details (optional)

I expect the ligature to be a merger of the `??` and `?=` ligatures.

### LINKED ISSUE #324: Feature Request: Support U+211E PRESCRIPTION INTAKE ℞
# Description of the new feature/enhancement (with images if possible)

It'd just be the R glyph with a stroke through.

### LINKED ISSUE #327: Feature Request: Ligature `<:>`
# Description of the new feature/enhancement (with images if possible)

A ligature for `<:>`.
Currently, the editor will display `<` and then the `:>` ligature, which looks lopsided and wrong.

### LINKED ISSUE #329: Font shows in mac office (Outlook/Word) as garbled characters
# Environment

```none
Cascadia Code version number: 2007.01
Application (with version) used to display text: mac Outlook 16.38, mac Word 16.38
OS platform and version: catalina 10.15.5
Screen resolution (i.e. 220dpi): 4k HiDPI at 200% scaling (1920x1080@2x)

Any other software? These fonts work fine in mac Excel. Same version (16.38)
```

# Steps to reproduce

Open a new word document, or new email, change font to any of the Cascadia family (code/mono/pl/..), then start typing. garbled chars start to show up.

The names of the fonts from the drop-down list in Outlook or Word are also showing as garbled fonts. They are not recognizable but you know they are the 4 Cascadia fonts installed.

# Expected behavior

Should not expect garbled chars show up.

# Actual behavior

garbled font name shown in the drop-down list. when selecting one of them, the fonts in main document will become garbled texts as well. 

screenshot1
screenshot2

### LINKED ISSUE #359: Missing character from IBM Graphics set.
Cascadia Code has the IBM Graphics characters found in positions x01–x1F of DOS codepages 437, 850, etc. But there is also a graphics character at x7F, Unicode U+2302 HOUSE, that is missing from Cascadia Code. I think that character is necessary for a terminal font.

Untitled

### LINKED ISSUE #371: NerdFont pacher height change
I know it's not your job, but please can you guys help me identify what is missing in the patched NerdFonts since 2008 onwards so that the height changes so significantly - ryanoasis/nerd-fonts#519?

I really like to use the new versions and I really like oh-my-zsh, but I am stuck at the 2005 version.

I'm attaching patched fonts from the latest version (where the result is the same). [static.zip](https://github.com/microsoft/cascadia-code/files/5290343/static.zip)

_Edit: added version_
Windows Terminal Preview
Version: 1.4.2652.0

### LINKED ISSUE #375: 【Issue Report】Difficult to identify quotation marks 【U+201C】【U+201D】【U+2018】【U+2019】
Up to 2009.22, this issue still exists.

It is difficult to distinguish between 【U+201C】`“` and【U+201D】`”`, 【U+2018】`‘` and【U+2019】`’`.

In some cases, I must recognize and use them correctly, so it is currently a little inconvenient so far.

Actually, many fonts have this problem, but I hope cascadia can improve.

### LINKED ISSUE #377: Equals and other horizontal characters uneven rendering on Windows 10 / VS Code
# Environment

```none
Cascadia Code version number: 2009.021
Application (with version) used to display text: Visual Studio Code 1.49.3
OS platform and version: Windows_NT x64 10.0.19041
Screen resolution (i.e. 220dpi): 3440x1440p @ 34' 100% scaling, but changing scaling does not help much 
```
Any other software?
```
Windows Terminal
IntelliJ
```

# Steps to reproduce

just type
```
=
```
or
```
===
```

# Expected behavior

I expected both lines to be the same width/height depends how you look at it also happens with other characters/ligatures; I can provide more examples if needed

The example below might look funny because its so enhanced but it is really noticeable when looking at text with normal zoom levels

# Actual behavior
3440x1440 34' screen

2560x1440 27' screen

and in comparison to Jetbrains Mono

### LINKED ISSUE #381: ᶿ thetamod U+1DBF should be taller
Regarding #357. The current thetamod ᶿ U+1DBF looks like obarredmod ᶱ U+1DB1.
Note: Glyphs.app GlyphData.xml calls ᶱ U+1DB1 barredomod instead of obarredmod, even though it calls ɵ U+0275 obarred.

The thetamod ᶿ U+1DBF should be vertically taller than obarredmod ᶱ U+1DB1.

If obarredmod ends up in the font, maybe consider adding its uppercase and lowercase equivalents: Obarred U+019F and obarred ɵ U+0275. Note: Again Glyphs app names them inconsistently: U+019F Ocenteredtilde, U+0275 obarred, U+1DB1 barredomod because their Unicode character names are not consistent either.
If added:
- Obarred U+019F can be a composite of Obarred-cy Ө U+04E8 (which is very slightly different from Theta Θ U+0398, a less than 5-units shift with the bar, maybe both Obarred and Obarred-cy should be composites of Theta)
- obarred ɵ U+0275 can be a composite of obarred-cy ө U+04E9.
- obarredmod ᶱ U+1DB1 is the current thetamod ᶿ U+1DBF.

# Environment

```none
Cascadia Code version number: 2009.22
Application (with version) used to display text: any
OS platform and version: any
Screen resolution (i.e. 220dpi): any

Any other software?
```

# Steps to reproduce
Use thetamod ᶿ for example in the test string "wʷθᶿ" or in the Comox/Sliammon "[gagat̓ᶿakʷup](https://www.firstvoices.com/explore/FV/sections/Data/Salish/Northern%20Salishan/Sliammon/learn/words/f9e32b18-8225-42b9-ba44-d0356109d009)".

# Expected behavior

The thetamod ᶿ should be taller than x-height modifier letters instead of the same vertical size.
# Actual behavior

The thetamod looks like an obarredmod ᶱ U+1DB1, a superscript obarred ɵ, instead of a superscript theta θ.

Cascadia-thetamod
gagat̓ᶿakʷup

### LINKED ISSUE #382: Fit nhookleft in advance width
The current design of nhookleft U+0272 ɲ extends left of the left-side-bearing. This means it may clash with some glyphs on the left.

# Description of the new feature/enhancement (with images if possible)

Change the design of nhookleft to fit within the 1200-unit advance width, either reducing width of the whole glyph or reducing the width of the n-part.

Here’s 
image
image
image

### LINKED ISSUE #383: esh ʃ, glottalstop ʔ, glottalstopreversed ʕ should be taller
esh U+0283 ʃ, glottalstop ʔ U+0294, glottalstopreversed U+0295 ʕ should be taller. They should match other round-ascender glyphs like f.

# Description of the new feature/enhancement (with images if possible)
- esh ʃ is currently a composite glyph made of integral, but integral is shorter than f (except in one weight).
- glottalstop ʔ and glottalstopreversed ʕ are both as tall as question but should be as tall a round ascender letter, like f.

fʃfʔfʕ

### LINKED ISSUE #384: Parentheses are asymmetric
# Description of the new feature/enhancement (with images if possible)
The parentheses are not symmetric across the x axis. I know this is trivial but, I'm sorry in my opinion it is very unpleasant and distracting to see. When paired next to each other as shown here, all I see is an incomplete heart shape.
image

Here's a screenshot of Fira Code's symmetric parentheses:
image
Ahh, much cleaner!

I think a font primarily designed for coding should really nail these common cases and this is a pretty painful one to see.

### LINKED ISSUE #386: Incomplete soft-dotting
# Environment

```none
Cascadia Code version number: 2009.22
Application (with version) used to display text: hb-view 2.7.2
OS platform and version: macOS 10.14.6
```

# Steps to reproduce

Render U+0456 CYRILLIC SMALL LETTER BYELORUSSIAN-UKRAINIAN I, U+0458 CYRILLIC SMALL LETTER JE, or U+1ECB LATIN SMALL LETTER I WITH DOT BELOW followed by an above-base combining mark.

# Expected behavior

These three characters all have the `Soft_Dotted` Unicode property, so the dot should disappear before an above-base combining mark, as it does for U+0069 LATIN SMALL LETTER I.

# Actual behavior

The soft dot does not disappear. Here is <U+0456, U+030A, U+0458, U+030A, U+1ECB, U+030A>:
і̊ј̊ị̊

### LINKED ISSUE #392: Suggestion: Widen angle quotation marks to visually contrast with less-than/greater-than signs
# Description of the new feature/enhancement
Angle quotation marks, or guillemets, are used in many languages to denote speech. They are also often used in a technical context as a means of demarcating email addresses (e.g., ‹notifications@github.com›) or denoting objects such as Unicode codepoints (e.g., ‹U+211D› DOUBLE-STRUCK CAPITAL R).

In Cascadia Code’s current design, the angle used in both the single and double angle quotation marks is quite narrow, relative to comparable programming and traditional fonts, and importantly, makes it more difficult to distinguish them from less-than/greater-than signs. For this reason, I’d like to propose **increasing/widening the angle for the following four codepoints**:
| Codepoint | Name |
| --- | --- |
| ‹U+2039› | SINGLE LEFT-POINTING ANGLE QUOTATION MARK |
| ‹U+203A› | SINGLE RIGHT-POINTING ANGLE QUOTATION MARK |
| ‹U+00AB› | LEFT-POINTING DOUBLE ANGLE QUOTATION MARK |
| ‹U+00BB› | RIGHT-POINTING DOUBLE ANGLE QUOTATION MARK |

## Examples, comparison with other fonts
_Render from LuaLaTeX, viewed in Evince PDF viewer—note the spacing and angle of Cascadia Code’s angle quotes in comparison to less-than/greater-than_
image

| LT/GT Sign | Single Angle Quote | Double Angle Quote | Angle Bracket |
| --- | --- | --- | --- |
| ‹U+003C› LESS-THAN SIGN    | ‹U+2039› SINGLE LEFT-POINTING ANGLE QUOTATION MARK  | ‹U+00AB› LEFT-POINTING DOUBLE ANGLE QUOTATION MARK  | ‹U+2329› LEFT-POINTING ANGLE BRACKET  |
| ‹U+003E› GREATER-THAN SIGN | ‹U+203A› SINGLE RIGHT-POINTING ANGLE QUOTATION MARK | ‹U+00BB› RIGHT-POINTING DOUBLE ANGLE QUOTATION MARK | ‹U+232A› RIGHT-POINTING ANGLE BRACKET |

_Angle brackets, less-than/greater-than signs and single guillemets in fonts Cambria, DejaVu Serif, Andron Mega Corpus, Andika and Everson Mono_ ([from Wikipedia](https://en.wikipedia.org/wiki/Guillemet#Shape))
image

# Proposed technical implementation details
Increase/widen the angle used in the design of both the single and double angle quotation marks to aid in visual differentiation from the less-than/greater-than signs, especially at small sizes.

### LINKED ISSUE #394: Bug Report (Lua inequality ~= rendered counter-intuitive)
# Environment

```none
Cascadia Code version number: 2009.021
Application (with version) used to display text: Windows Terminal 1.5.3242.0
OS platform and version: Windows 10 10.0.19041.630
Screen resolution (i.e. 220dpi): 

Any other software?
```

# Steps to reproduce

This is follow-up to the FiraCode report https://github.com/tonsky/FiraCode/issues/1171 which appeared to be already addressed in their modern releases. Some time ago they have ligature ~= which rendered confusingly as approximately equal. There is no language out there, where this ligature is correct, and there is language where it's shown counter-intuitively - Lua uses ~= as not equal.

# Expected behavior

For consistency reasons I'd prefer you to delete this ligature from Cascadia Code similar as they have deleted it from Fira Code.
https://github.com/tonsky/FiraCode/issues/381

# Actual behavior

image

### LINKED ISSUE #395: Underline position and thickness metrics are not correct
# Environment

```none
Cascadia Code version number: 2009.22
Application (with version) used to display text: Windows Terminal 1.4.3243.0
OS platform and version: Windows 10
Screen resolution (i.e. 220dpi):  -
```

# Steps to reproduce

Inspect underline thickness and position with larger font sizes.

# Expected behavior

I expect underline to be more thick with larger font sizes (if you compare to other fonts).

# Actual behavior

Seems like underline thickness is always reported around `1px`. I suspect that underline position metric is also off (should probably be offset by more pixels with larger font sizes, as right now this seem to be constant value no matter the size).

Screenshot showing the issue in Windows Terminal (with font size set to `40px`):

underline-metrics-issue

### LINKED ISSUE #400: Better differentiate ° degree from º ordmasculine
# Description of the new feature/enhancement
The ° degree and º ordmasculine glyphs are very similar and hard to distinguish at most sizes.

Here are C°Cºªoa  in weight=200,400,700, that's /C/degree/C/ordmasculine/ordfeminine/o/a:
Cascadia-Code-700-C°Cºªoa-16
Cascadia-Code-400-C°Cºªoa-16
Cascadia-Code-200-C°Cºªoa-16

# Proposed technical implementation details (optional)
Some possible changes to /degree:
- make is smaller
- make it have a thinner stroke especially in normal and bolder weights.
- make it sit at a different height from ordmasculine
Possible change to /ordmasculine:
- make it less round (although this one is hard with the proportions of other superscripts).

### PR DESCRIPTION
**Pull request by aaronbell**
This is a significant update to Cascadia Code including a large number
of bug fixes as well as updating the font to offer support for Fira
Code v5 ligature support. 

This update supersedes PR #373.

Closes #262 - ⏎ added
Closes #264 - additional codepoints for control characters added
Closes #281 - `!:` and `!.` added
Closes #290 - `/\` and `\/` added
Closes #301 - `??=` added
Closes #324 - ℞ added
Closes #327 - `<:>` and other variants implemented via the `calt`
  refactoring
Closes #359 - house added
Closes #371 - Added x-height instruction into ttfautohint to control the
  height of the lowercase.  
Closes #375 - Completely redesigned quote marks for better recognition
Closes #377 - updated hinting to achieve more consistent results
Closes #381 - increased height of thetamod
Closes #382 - reduced the width of the hooklefts
Closes #383 - updated heights on esh, glottalstop, glottalstopreversed
Closes #384 - tweaked hinting a little bit. Maybe it'll help :)
Closes #386 - added remaining soft-dotting
Closes #392 - changed designs of angled quotes (they are now round)
Closes #394 - changed former `~=` symbol to a simpler component-based
  version. Should be less confusing now for Lua / Matlab users. 
Closes #395 - makes the underline thicker based on font weight
Closes #400 - increased size of degree

Closes #219 
The full control pictures block has been added (u+2400 to u+2426). For
purposes of rendering, the two letter abbreviations have been used
instead of the standard three letter abbreviations:

Additionally, ss20 includes the oft-unused graphical representations of
these codepoints (for fun!):

Closes #276 (infinite arrows)
Full support for Fira Code's current ligature set (with a few
exceptions). Now featuring infinite arrows!!! 

This involved a full refactoring of the `calt` feature—for those
interested, it now uses forward-looking substitutions instead of
backward-looking substitutions and progressive substitution to reduce
code. This also required some redesigning of the greater / lesser
related ligatures. Please note, I have also removed all the obsolete
ligatures now covered by the arrows code.

Closes #329 
There was a mismatch in the font's postscript naming conventions that
was corrected. Should now render all weights in Word. **Note** there is
apparently an additional bug in Mac Word's implementation of variable
fonts which should be available in an update mid-Feb. 

* Not listed – Reworked the hints for the mod and superscript glyphs so
  that they're bottom-up rather than top-down. This allows for better
  bottom alignments. 

Aside from the above changes, this version also includes many other
small updates including spacing, outline quality improvements, and
fixing hinting.

### CODE REVIEW

**Comment by DHowett** [comment]
Wow, it's lovely!

What do the non-smart quotes look like? Boring old U+0022 and U+0027?

**Comment by DHowett** [comment]
(I'm concerned about the conflicts in the `.fea` files from other updates -- looks like there's something missing in this PR?)

**Comment by aaronbell** [comment]
> What do the non-smart quotes look like? Boring old U+0022 and U+0027?

Yup. Boring old. No need to change those fellows!

**Comment by aaronbell** [comment]
> (I'm concerned about the conflicts in the `.fea` files from other updates -- looks like there's something missing in this PR?)

I took a look at the feature code and, aside from the massive restructuring of the calt feature, everything else looks pretty similar. Is there something you're particularly concerned about?

**Comment by DHowett** [comment]
> > (I'm concerned about the conflicts in the `.fea` files from other updates -- looks like there's something missing in this PR?)
> 
> I took a look at the feature code and, aside from the massive restructuring of the calt feature, everything else looks pretty similar. Is there something you're particularly concerned about?

Nope, just the worry every time I see a conflict :smile: Thanks!

**Comment by aaronbell** [comment]
Alright, I think this is ready to go for this release.

**Comment by schuelermine** [comment]
Oooh, I'm excited to download this!

File: sources/CascadiaCode-Bold.ufo/fontinfo.plist [hunk 1]
```diff
@@ -204,7 +204,7 @@
     <key>openTypeNameLicense</key>
     <string>Microsoft supplied font. You may use this font to create, display, and print content as permitted by the license terms or terms of use, of the Microsoft product, service, or content in which this font was included. You may only (i) embed this font in content as permitted by the embedding restrictions included in this font; and (ii) temporarily download this font to a printer or other output device to help print content. Any other use is prohibited.
  
-The following license, based on the SIL Open Font license (https://scripts.sil.org/OFL), applies to this font. Additional license terms may be found on the GitHub repository for this font (https://github.com/microsoft/cascadia-code/blob/main/LICENSE).
+The following license, based on the SIL Open Font license (https://scripts.sil.org/OFL), applies to this font. Additional license terms may be found on the GitHub repository for this font (https://github.com/microsoft/cascadia-code/blob/master/LICENSE).
```
Comment (DHowett) [inline sources/CascadiaCode-Bold.ufo/fontinfo.plist:207]: This merge is incorrect -- it changed the URL back from `main` to `master`!

Comment (aaronbell) [reply sources/CascadiaCode-Bold.ufo/fontinfo.plist:207]: Oh fun. Must have been an old version on my other computer that snuck in.

**Comment by DHowett** [comment]
image

It's our intent that all runs of straight lines ligate now, right? Neat!

**Comment by aaronbell** [comment]
Yup! So you can make infinitely long arrows or other bits. Try out using `>` `|` or `<` in conjunction :)

**Comment by aaronbell** [comment]
Yup! So you can make infinitely long arrows or other bits. Try out using `>` `|` or `<` in conjunction :)

**Comment by DHowett** [comment]
In that case, _let's do it_!. I'll merge.

**Comment by DHowett** [comment]
Screen Shot 2021-02-02 at 3 24 00 PM
Screen Shot 2021-02-02 at 3 24 05 PM
Screen Shot 2021-02-02 at 3 33 54 PM
Screen Shot 2021-02-02 at 3 35 37 PM
Screen Shot 2021-02-02 at 3 40 39 PM
Screen Shot 2021-02-02 at 3 41 52 PM
Screen Shot 2021-02-02 at 3 43 09 PM
Screen Shot 2021-02-02 at 3 44 41 PM
Screen Shot 2021-02-02 at 3 50 01 PM
Screen Shot 2021-02-02 at 3 47 53 PM
Screen Shot 2021-02-02 at 4 06 07 PM
Screen Shot 2021-02-02 at 3 54 18 PM

**Comment by aaronbell** [comment]
_the joy of closing 23 bugs_

**Comment by DHowett** [comment]
Well done! Thanks for all the excellent, gorgeous fixes. I've got it in build now, and I'll kick the tires a bit before including it in Terminal's upcoming bugfix releases :)

**Comment by DHowett** [comment]
image


This may throw some people off, but I do get why it is the way it is!

**Comment by aaronbell** [comment]
Yeah, understandable. This is following Fira Code, though, which has had this sort of set up for a while. So if folks aren’t angry over there, I’m ok with it

**Comment by schuelermine** [comment]
Is this going to be released as a GitHub release? The last one is from September last year.

**Comment by DHowett** [comment]
@schuelermine no we were going to force every person to compile the font from source for the rest of eternity

_yes, of course we're going to release it as a GitHub releas.e_

**Pull request merged**